### ISSUE TITLE
EPUBCFIs don't validate

### ISSUE THREAD

**Issue by mattgarrish** [w3c/epubcheck#1405]
I just spotted that brackets in CFIs are being reported as errors now. The georgia-cfi test book now emits messages like these:

```
ERROR(RSC-020): .\30\georgia-cfi.epub/EPUB/nav.xhtml(-1,-1): "package.opf#epubcfi(/6/4[ct]!/4/2[d10e42]/12[d10e85]/6[d10e93]/1:1552[Bryan,%20and])" is not a valid URI (Illegal character in fragment: ?[? is not allowed)
ERROR(RSC-020): .\30\georgia-cfi.epub/EPUB/nav.xhtml(-1,-1): "package.opf#epubcfi(/6/4[ct]!/4/2[d10e42]/12[d10e85]/6[d10e93]/1:1552[Bryan,%20and])" is not a valid URI (Illegal character in fragment: ?[? is not allowed)
ERROR(RSC-020): .\30\georgia-cfi.epub/EPUB/nav.xhtml(-1,-1): "package.opf#epubcfi(/6/4[ct]!/4/2[d10e42]/18[d10e150]/4[d10e155]/1:35)" is not a valid URI (Illegal character in fragment: ?[? is not allowed)
```

I guess, being its own fragment scheme, allowing brackets unescaped is okay, but it does seem to run counter to URL rules.

So while we no longer formally endorse CFIs in content, they are still allowed, so we'll need to figure out who wins here.

**Comment by mattgarrish** [comment]
Sorry for the busted opening of the issue. Accidentally tapped the mousepad at an inopportune spot.

**Comment by rdeltour** [comment]
So according the URL standard:
> A **URL-query string** must be zero or more [URL units](https://url.spec.whatwg.org/#url-units)

and
> The **URL units** are [URL code points](https://url.spec.whatwg.org/#url-code-points) and [percent-encoded bytes](https://url.spec.whatwg.org/#percent-encoded-byte).

and
> The **URL code points** are [ASCII alphanumeric](https://infra.spec.whatwg.org/#ascii-alphanumeric), U+0021 (!), U+0024 ($), U+0026 (&), U+0027 ('), U+0028 LEFT PARENTHESIS, U+0029 RIGHT PARENTHESIS, U+002A (*), U+002B (+), U+002C (,), U+002D (-), U+002E (.), U+002F (/), U+003A (:), U+003B (;), U+003D (=), U+003F (?), U+0040 (@), U+005F (_), U+007E (~), and [code points](https://infra.spec.whatwg.org/#code-point) in the range U+00A0 to U+10FFFD, inclusive, excluding [surrogates](https://infra.spec.whatwg.org/#surrogate) and [noncharacters](https://infra.spec.whatwg.org/#noncharacter).

So unexpectedly, it seems the Galimatias URL parser is right.

Let's move that to a spec issue? Or do you think EPUBCheck should willfully violate the spec to allow these CFIs?

**Comment by mattgarrish** [comment]
> Or do you think EPUBCheck should willfully violate the spec to allow these CFIs?

I've never heard that anyone uses CFIs in content, so we're probably fine leaving the error in place and opening a spec issue about this. I don't want to make work for something that isn't proven to be implemented.

But as no one is working on CFIs in W3C, it's not going to get resolved any time soon, if ever.

**Comment by mattgarrish** [comment]
Actually, digging down into the CFI spec they acknowledge unsafe characters escaping in [example 6](https://w3c.github.io/epub-specs/epub33/epubcfi/#example-6), but suggest this is "aggressive".

So arguably it's the sample file that needs to be fixed here, but I'll still open an issue that the CFI spec may need to be more realistic about using unencoded brackets.

Feel free to close this as not needing any action.

**Comment by rdeltour** [comment]
Closing as `wontfix`. We can reopen if someone complains.
